backend/src/main/scala/bloop/io/ClasspathHasher.scala

```diff
@@ -66,219 +49,100 @@ object ClasspathHasher {
+       * downstream clients on the assumption that the hashing of this
+       * entry is too slow because of something that happened to this
+       * process. The completion of the downstream promise will also log a
+       * warning to the downstream users so that they know that a hashing
+       * process is unusually slow.
+       */
+      val timeoutFallback: Task[FileHash] = Task {
+        val msg =
+          s"Hashing ${path} is taking more than ${timeoutSeconds}s"
+        try {
+          logger.warn(msg)
+          serverOut.println(msg)
+        } catch { case NonFatal(_) => () }
+        BloopStamps.emptyHash(path)
       }
-    }
-  }
 
-  private[this] val definedMacrosJarCache = new ConcurrentHashMap[File, (JarMetadata, Boolean)]()
-
-  private val blackboxReference = "scala/reflect/macros/blackbox/Context".getBytes
-  private val whiteboxReference = "scala/reflect/macros/whitebox/Context".getBytes
-  def containsMacroDefinition(classpath: Seq[File]): Task[Seq[(File, Boolean)]] = {
-    import org.zeroturnaround.zip.commons.IOUtils
-    import org.zeroturnaround.zip.{ZipEntryCallback, ZipUtil}
-    def readJar(jar: File): Task[(File, Boolean)] = Task {
-      if (!jar.exists()) sys.error(s"File ${jar} doesn't exist")
-      else {
-        def detectMacro(jar: File): Boolean = {
-          var found: Boolean = false
-          ZipUtil.iterate(
-            jar,
-            new ZipEntryCallback {
-              override def process(in: InputStream, zipEntry: ZipEntry): Unit = {
-                if (found) ()
-                else if (zipEntry.isDirectory) ()
-                else if (!zipEntry.getName.endsWith(".class")) ()
-                else {
-                  try {
-                    val bytes = IOUtils.toByteArray(in)
-                    found = {
-                      bytes.containsSlice(blackboxReference) ||
-                      bytes.containsSlice(whiteboxReference)
-                    }
-                  } catch {
-                    case t: Throwable => println(s"Error in ${t}")
-                  }
-                }
-              }
-            }
-          )
-          found
-        }
-
-        val attrs = Files.readAttributes(jar.toPath, classOf[BasicFileAttributes])
-        val currentMetadata = (FileTime.fromMillis(IO.getModifiedTimeOrZero(jar)), attrs.size())
+      hash(path).timeoutTo(
+        duration = timeoutSeconds.seconds,
+        backup = timeoutFallback
+      )
+    }
 
-        Option(definedMacrosJarCache.get(jar)) match {
-          case Some((metadata, hit)) if metadata == currentMetadata => jar -> hit
-          case _ =>
-            val detected = detectMacro(jar)
-            definedMacrosJarCache.put(jar, (currentMetadata, detected))
-            jar -> detected
+    val onCancel = Task(isCancelled.compareAndSet(false, true)).void
+
+    val tasks = classpath.toVector.map(path => makeHashingTask(path.underlying))
```

Comment:
core logic is pretty much the same, but expressed much simpler. Instead of creating array with nulls and mutable listBuffers, current logic just creates list of tasks and then sequence them.